rfcs/62-editor-data-model.md

```diff
@@ -0,0 +1,562 @@
+
+In contrast, we want to enable the Editor to observe changes in the Game code without rebuilding or even restarting the Editor process, in order to enable fast iteration for the Authors. This feature, referred to as _hot-reloading_, requires the Editor to support type changes, which is not possible with static types like the Game uses. Therefore, to allow this flexibility, the Editor data model is based on _dynamic types_ defined exclusively during Editor execution. The data model defines an API to instantiate objects from dynamic types, manipulates the data of those instances, but also modify the types themselves while objects are instantiated. This latter case is made possible through a set of migration rules allowing to transform an object instance from a version of a type to another modified version of that type.
+
+The general workflow is as follows:
+
+- The Author starts the Editor executable and create a new project or loads an existing project.
+- The Editor launches the Game executable in _edit mode_, a special mode whereby the `EditorClientPlugin` is enabled and allows the Editor and the Game to communicate in real time. The `EditorClientPlugin` is a Bevy plugin (implements the `Plugin` trait) provided by an Editor crate and used during development only, which implements the Game side of the Editor/Game communication. It's compiled out of any Release build before shipping the Game.
+- The Editor queries the Game for all its custom types, including components and resources.
+- The Editor follows the same process for any third-party library, to enable extending Bevy with new functionalities.
+- The Editor builds a database of all (dynamic) types, merging the built-in Bevy types (the ones defined in any `bevy_*` official crate), any third-party library, and the Game.
+- The Author creates new object instances from any of those types, and manipulates the data of those instances by setting the value of the properties exposed by the type of the object.
+- Optionally, the Author saves the project, which makes the Editor serialize the data model to persistent storage for later Editor sessions on the same project.
+
+The data model supports various intrinsic types: boolean, integral, floating-point, strings, arrays, dictionaries, enums and flags, and objects (references to other types). Objects can be references to entities or to components. Dynamic types extend those by composition; they define a series of _properties_, defined by their name and a type themselves. Conceptually, a property can look like (illustrative pseudo-code):
+
+```txt
+type "MyPlayerComponent" {
+  property "name"  : string
+  property "health": float32
+  property "target": entity
+  property "weapon": type "MyWeaponComponent"

```

Comment:
Those two properties are objects, right? One referencing an entity, and the other referencing a component. I'm wondering if comments may help others understanding that or if it only adds noise.

Reply:
I have to re-work that part, this is ill-defined because the first one is indeed a _reference_ to another object (via its `Entity`), but the second one it's unclear if it's a reference to another component or if it's an aggregation (struct as field inside a struct).

Reply:
I've reworked that pseudo-code, I think it's more clear. Let me know @Nilirad if that's better.